transform.py

```diff
@@ -0,0 +1,169 @@
+from dotenv import load_dotenv
+from openai import OpenAI
+import tiktoken
+import prompts
+from extract import get_listing_data
+
+
+load_dotenv()
+client = OpenAI(api_key=getenv("OPENAI_API_KEY"))
+
+
+def shorten_text_by_tokens(
+    text, keep_start_tokens=4000, keep_end_tokens=4000, placeholder="[...]"
+):
+    """Shortens court transcript by removing tokens from the middle"""
+    encoder = tiktoken.encoding_for_model("gpt-4o-mini")
+    tokens = encoder.encode(text)
+
+    if len(tokens) <= keep_start_tokens + keep_end_tokens:
+        return text
+
+    start_tokens = tokens[:keep_start_tokens]
+    end_tokens = tokens[-keep_end_tokens:]
+
+    start_text = encoder.decode(start_tokens)
+    end_text = encoder.decode(end_tokens)
+
+    shortened_text = f"{start_text}{placeholder}{end_text}"
+
+    return shortened_text
+
+
+def get_summary(prompt: str, transcript: str):
+    """Collect data about the transcript using the GPT-4o-mini model"""
+    completion = client.chat.completions.create(
+        model="gpt-4o-mini",
+        messages=[
+            {"role": "system", "content": prompt},
+            {"role": "user", "content": transcript},
+        ],
+        temperature=0.1,
+    )
+    print(completion.usage)
+    return completion
+
+
+def validate_gpt_response(data: dict) -> bool:
+    """Validates the data shape extracted from the GPT-4o-mini API"""
+    expected_keys = {
+        "verdict": str,
+        "summary": str,
+        "case_number": str,
+        "verdict_summary": str,
+        "judge": list,
+        "tags": list,
+        "first_side": dict,
+        "second_side": dict,
+    }
+
+    for key, expected_type in expected_keys.items():
+        value = data.get(key)
+        if value is None:
+            return False
+        if not isinstance(value, expected_type):
+            return False
+        if isinstance(value, str) and not value.strip():
+            return False
+        if isinstance(value, list) and len(value) == 0:
+            return False
+    return True
+
+
+def get_data(html_data: list[dict], index: int) -> dict:
+    """Combines the html data with the GPT-4o-mini data for a single case"""
+    transcript = html_data[index].get("text_raw")
+    shortened_transcript = shorten_text_by_tokens(transcript)
+    del html_data[index]["text_raw"]
+    user_message = prompts.user_message + shortened_transcript
+    try:
+        gpt_data = literal_eval(
+            get_summary(prompts.system_message, user_message).choices[0].message.content
+        )
+    except (SyntaxError, ValueError) as e:
+        return None
+    data = html_data[index] | gpt_data
+    return data
+
+
+def format_date(date_string: str) -> date:
+    """Converts a date string to a date object"""
+    if not date_string:
+        return None
+    clean_date_string = date_string.replace(", midnight", "")
+    return datetime.strptime(clean_date_string, "%d %b %Y").date()
+
+
+def convert_dict_to_tuple(data):
```

Comment:
return typehint?

Reply:
Added input and return type hints, thanks Anna